README.md
# Consul-Alerts Cookbook
This cookbook provides an LWRP as well as a sample recipe for managing
[consul-alerts](https://github.com/AcalephStorage/consul-alerts)

[![Build Status](https://api.shippable.com/projects/54597b1aa85d45d063d9033a/badge?branchName=master)](https://app.shippable.com/projects/54597b1aa85d45d063d9033a/builds/latest)


## Requirements
* Chef 11+
* A working [Consul](https://consul.io/) installation

## Supported Platforms
The following is a list of platforms that the cookbook is regularly tested on.
It is likely it runs on other platform combinations as well, but only those
that are regularly tested on will be declared *supported*:

* EL6 (tested using Centos)
* EL5 (tested using Centos)

## Usage
The simplest way to get started is to add `recipe[consul-alerts]` to a node's
run_list that already has consul installed. This will invoke the default
recipe which simply calls the LWRP with a very minimal set of options. Odd's
are that you environment is a little more complex, and you'll likely want to
write a wrapper cookbook/recipe that invokes the LWRP with settings appropriate
for your environment

## LWRP
The cookbook provides a `consul_alert` LWRP which is used to install and manage
the `consul-alerts` daemon

### Actions
* `:create` - This will install the necessary components to run the daemon.
* `:remove` - This will remove any components that were installed by the LWRP.
* `:start` - This will start the daemon on the host. It will also be configured
  to start on boot.
* `:stop` - This will stop the daemon on the host. It will also **disable**
  start on boot.

### Attributes
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>version</tt></td>
    <td>String</td>
    <td>Which version of `consul-alerts` to install</td>
    <td><tt>0.1.1</tt></td>
  </tr>
  <tr>
    <td><tt>install_dir</tt></td>
    <td>String</td>
    <td>The directory into which `consul-alerts` will be installed</td>
    <td><tt>/usr/local/consul-alerts</tt></td>
  </tr>
  <tr>
    <td><tt>user</tt></td>
    <td>String</td>
    <td>The user account under which the daemon will be run</td>
    <td><tt>consul-alerts</tt></td>
  </tr>
  <tr>
    <td><tt>group</tt></td>
    <td>String</td>
    <td>The group under which the daemon will be run</td>
    <td><tt>consul-alerts</tt></td>
  </tr>
  <tr>
    <td><tt>service_name</tt></td>
    <td>String</td>
    <td>The name of the service to give the daemon</td>
    <td><tt>consul-alerts</tt></td>
  </tr>
  <tr>
    <td><tt>consul_addr</tt></td>
    <td>String</td>
    <td>The address of the consul HTTP API</td>
    <td><tt>127.0.0.1:8500</tt></td>
  </tr>  
  <tr>
    <td><tt>consul_dc</tt></td>
    <td>String</td>
    <td>The consul datacenter to use</td>
    <td><tt>dc1</tt></td>
  </tr>
  <tr>
    <td><tt>port</tt></td>
    <td>Fixnum</td>
    <td>The TCP port on which the daemon will listen</td>
    <td><tt>9000</tt></td>
  </tr>
  <tr>
    <td><tt>base_url</tt></td>
    <td>String</td>
    <td>The base URL from which `consul-alerts` can be downloaded</td>
    <td><tt>http://dl.bintray.com/darkcrux/generic</tt></td>
  </tr>
  <tr>
    <td><tt>checksums</tt></td>
    <td>Hash</td>
    <td>A Hash of known Checksums for the various downloadable files.
      It is unlikely you need (or want) to adjust this, however if for some 
      reason you need to override, or add additional checksums, you can do so 
      here.
    </td>
    <td>Refer to the `resource/default.rb`</td>
  </tr>
  <tr>
    <td><tt>config</tt></td>
    <td>Hash</td>
    <td>A Hash of configuration key/value pairs that will be inserted into the
      Consul Key/Value store.
    </td>
    <td><tt>
      'checks/enabled' => 'true', <br />
      'checks/change-threshold' => '60', <br />
      'events/enabled' => 'true', <br />
      'notifiers/log/enabled' => 'true', <br />
      'notifiers/log/path' => '/var/log/consul-notifications.log' <br /> 
    </tt></td>
  </tr>
</table>

## Contributing
1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Testing

This cookbook currently uses [test-kitchen](https://github.com/opscode/test-kitchen)
along with [ChefSpec](https://docs.getchef.com/chefspec.html).

Test Kitchen: `kitchen test`

ChefSpec: `bundle exec rspec`

## Releasing
This cookbook uses an 'even number' release strategy. The version number in master
will always be an odd number indicating development, and an even number will
be used when an official build is released.

Come release time here is the checklist:

* Ensure the `metadata.rb` reflects the proper *even* numbered release
* Ensure there is a *dated* change log entry in `CHANGELOG.md`
* Commit all the changes
* Use stove to release (`bundle exec stove`)
* Bump the version in metadata.rb to the next *patch level* odd number

## Authors
- Author: David Petzel (<davidpetzel@gmail.com>)



CHANGELOG.md
# Consul-Alerts Cookbook Change Log

## v1.0.0 (Nov 9, 2014)
* Initial Release


spec/spec_helper.rb
require 'coveralls'
Coveralls.wear!

require 'chefspec'
require 'chefspec/berkshelf'
require_relative 'support/matchers'

ChefSpec::Coverage.start!

RSpec.configure do |config|
  config.platform = 'redhat'
  config.version = '6.5'
end


spec/support/matchers.rb
def create_service_factory(resource_name)
  ChefSpec::Matchers::ResourceMatcher.new(:service_factory, :create, resource_name)
end


spec/lwrp/default_spec.rb
require_relative '../spec_helper'

describe 'consul_alerts::default' do
  let(:chef_run) do
    ChefSpec::ServerRunner.new(step_into: ['consul_alerts']) do |node|
      node.set[:consul_alerts][:version] = '0.1.1'
    end.converge(described_recipe)
  end

  it 'does something' do
    expect(chef_run).to create_remote_file('consul-alerts-tar').with(path:
      ::File.join(Chef::Config[:file_cache_path], 'consul-alerts-0.1.1-linux-amd64.tar'))
  end
end


resources/default.rb
actions :create, :remove, :start, :stop
default_action :create

attribute :version,
          kind_of: String,
          default: '0.1.1'

attribute :install_dir,
          kind_of: String,
          default: '/usr/local/consul-alerts'

attribute :user,
          kind_of: String,
          default: 'consul-alerts'

# The group which the consul-alerts daemon will run as
attribute :group,
          kind_of: String,
          default: 'consul-alerts'

attribute :service_name,
          kind_of: String,
          default: 'consul-alerts'

attribute :consul_addr,
          kind_of: String,
          default: '127.0.0.1:8500'

attribute :consul_dc,
          kind_of: String,
          default: 'dc1'

attribute :port,
          kind_of: Fixnum,
          default: 9000

attribute :bind_addr,
          kind_of: String,
          default: '0.0.0.0'

attribute :base_url,
          kind_of: String,
          default: 'http://dl.bintray.com/darkcrux/generic'

attribute :checksums,
          kind_of: Hash,
          default: {
            'consul-alerts-0.1.0-darwin-amd64.tar' =>
              'c8741ca8ec843a663c34a26f038e32ee25b89b0e063a08e1b234870dff923ab1',
            'consul-alerts-0.1.0-linux-386.tar' =>
              '4253b3d091f4810e938fc79715254c0498bb71fcbc70156f1ea4a7b371d77e86',
            'consul-alerts-0.1.0-linux-amd64.tar' =>
              '438ac92588affa856069d7b117f7aee9528b8319b27e54b969d4cd6ef25c1cc0',
            'consul-alerts-0.1.1-darwin-amd64.tar' =>
              'a936bc4c71a7768ceec40ddb4539fe3332bcf2e1cba61d9ef867911780a5a59b',
            'consul-alerts-0.1.1-linux-386.tar' =>
              '4e3e697ead0509d04fb43e5056692bab8d037f706ad682e64aba8083dcabdec4',
            'consul-alerts-0.1.1-linux-amd64.tar' =>
              'c6a4a39f4c17770bd2d9b552e0baad07fe8c8a7d153ec29efcad41b595210a62'
          }
attribute :config,
          kind_of: Hash,
          default: {
            'checks/enabled' => 'true',
            'checks/change-threshold' => '60',
            'events/enabled' => 'true',
            'notifiers/log/enabled' => 'true',
            'notifiers/log/path' => '/var/log/consul-notifications.log'
          }

attr_accessor :tarfile_url, :tarfile_name

def tarfile_name
  "consul-alerts-#{version}-#{node[:os]}-#{arch}.tar"
end

def arch
  node[:kernel][:machine] =~ /x86_64/ ? 'amd64' : '386'
end

def tarfile_url
  "#{base_url}/#{tarfile_name}"
end

def tarfile_checksum
  checksums[tarfile_name]
end

def validate_platform!
  valid = false

  case node[:platform_family]
  when 'rhel'
    valid = valid_rhel_version?
  when 'debian'
    valid = valid_debian_version?
  end

  if valid == false
    Chef::Application.fatal!('consul-alerts LWRP not supported on' \
      " #{node[:platform]} #{node[:platform_version]}")
  end
  true
end

def validate_chef_version!
  return true if Chef::Version.new(Chef::VERSION).major >= 11
  Chef::Application.fatal!(
    'The consul-alerts LWRP is only supported on Chef 11 or newer')
end

private

def valid_rhel_version?
  valid = false
  major_ver = node[:platform_version].to_i
  valid = true if major_ver == 6 || major_ver == 5
  valid
end

def valid_debian_version?
  valid = false
  major_ver = node[:platform_version].to_i
  valid = true if node[:platform] == 'ubuntu' && major_ver >= 14
  valid
end
